### ISSUE TITLE
Grow pedestal and noise database

### ISSUE THREAD

**Issue by jrquirk** [alcap-org/AlcapDAQ#186]
Any interest in growing this to hold things like channel_marked_as_disabled and constant_fraction_time_offset and max_bin_time_offset?

**Comment by AndrewEdmonds11** [comment]
Yes, I think this is a good idea. Is it also worth adding it to the production.db database or is it best to keep them separate?

**Comment by benkrikler** [comment]
I was thinking that axis limits for plots might be useful in a database. For amplitudes it's fairly simple to just use the adc range but for time differences and integrals it's less straight forward, and would be nice to have controls over the axes without massive `if.. else` blocks to decide which channel we're on.

**Comment by AndrewEdmonds11** [comment]
Also it might nice to put in the parameter values for the PulseCandidateFinder (maybe on a dataset-by-dataset basis rather than run-by-run)

**Comment by jrquirk** [comment]
> Is it also worth adding it to the production.db database or is it best to keep them separate?

A similar issue has come up [before](https://github.com/alcap-org/AlcapDAQ/issues/104#issuecomment-46954320), and I'd tend to agree with @litchfld that keeping them separate has it's appeal. For me it's the fear of screwing up a single database being diminished if there's some separation.

**Comment by AndrewEdmonds11** [comment]
As I've just referenced in another issue. We should also add fields for whether we want to analyse a particular channel.

**Comment by jrquirk** [comment]
I would like to pick this up. @AndrewEdmonds11, is that cool with you? You're the only one I think I'd be clashing with, and I want to make sure I'm not reproducing work.

**Comment by AndrewEdmonds11** [comment]
Sure. I was only going to try and move some stuff that I have outside of the SetupNavigator inside. In particular:
- fill fOneSigmaValues and fPedestalValues in SetupNavigator (rather than PulseCandidateFinder) and read the values from there (this should also be better for everyone)
- also something else on my to-do list is to make it crash loudly and informatively

Shall I try and do this now and then you take over from my evening (your afternoon)?

**Comment by jrquirk** [comment]
That works for me. I'll keep an eye out for a published feature later today and merge anything onto my branch.

**Comment by AndrewEdmonds11** [comment]
Ok, John. I've just merged my work into develop.

Now we have:
- (almost) only one place where we have a `TSQLiteServer`
- cache the pedestals and noises from the database in `SetupNavigator`
- get the pedestals and noises from `SetupNavigator` using `IDs::channel`

There is one thing left and that is writing to the SQLite database. Currently, the way I've done it in PlotPedestalAndNoise is the only place we do it but it would be good to move this in to SetupNavigator as well. Especially things that we would do every time we write. i.e.:
- create the table if it doesn't exist
- choose between inserting and updating entries

I can't think of the best way make this common with all the possible things we might fill into a database (e.g. could we guarantee that everything will be run number, channel, bank name and then the value(s) we are storing?). 

Anyway, I'm not going to look into this now since I think it's too much for 5.30pm on a Monday but do feel free to do it as you're doing your stuff. I'll be in e-mail contact tonight if you have any questions/complaints about stuff I've done to the SetupNavigator.

**Comment by litchfld** [comment]
Seeing the boilerplate Andy was working on yesterday i'm thinking we ought to have a DB wrapper class that hides the process of reading and (particularly) cacheing tables in memory.  Seems like an amusing thing to work on, but i doubt i'll have time for a few weeks.  If i get bored enough of writing talks i might do it anyway.

**Comment by jrquirk** [comment]
Andy caches the values he needs in memory. I was going to try to follow the same pattern.

**Comment by benkrikler** [comment]
> There is one thing left and that is writing to the SQLite database. Currently, the way I've done it in PlotPedestalAndNoise is the only place we do it but it would be good to move this in to SetupNavigator as well.

It does sound like writing the database should be handled in SetupNavigator.  However we implement writing the databases though, can we keep in mind writing to the DB in parallel?  I couldn't find the handles for identifying a locked DB through Root's sqlite interface.  If there is no way, then perhaps we would need to have the SetupNavigator write to a file identified by the run number which we can later concatenate with other runs  to produce the input database.

Also, since the contents of this file is growing, we should probably rename it to something more generic than 'pedestals-and-noises.sqlite'.

**Comment by jrquirk** [comment]
I've had trouble with this, as in I've had trouble reproducing the bug. The python SQLite library has much better documentation than the ROOT one, so the best I can do is try to apply that logic.

I think before starting any write operation, we should call `TSQLServer::BeginTransaction()` and afterwords call `TSQLServer::Commit()`. Obviously we've been writing to the database without calling `BeginTransaction`, but a call to `Commit` should free up any write locks according to both the python docs and the sqlite [docs](http://www.sqlite.org/lockingv3.html). However, ROOT was giving me errors if I tried to commit with an error saying something like no transaction was started.

Since I haven't reproduced the issue, I can't tell if this will fix it, but this works in the python libs for the production db.

Does this sort of make sense?

**Comment by benkrikler** [comment]
>  I've had trouble reproducing the bug.

It occurred for me the second I used the PlotPedestalAndNoise module with the `export_sql` option and running some 10 or so of these jobs at once on the batch system. 

Perhaps the recent changes to DB access have resolved the problem then?  If one job had a write lock on the DB then that would have prevented read access to the DB which was needed for every request to the setup navigator (I think) before Andy implemented the `CacheCalibDB` method.

> The python SQLite library has much better documentation than the ROOT one.

If only ROOT was as well documented as rootana, eh?  

> I think before starting any write operation, we should call TSQLServer::BeginTransaction() and afterwords call TSQLServer::Commit()

When I was looking at the documentation I got the impression that the TSqliteServer didn't really implement the abstract TSQLServer class completely.  A few methods state they should be ignored or not used for an sqlite database.  Maybe the `BeginTransaction` and `Commit` methods fall under this category?

**Comment by litchfld** [comment]
I'm beginning to think this is more trouble than it's worth. I'm also more than a little nervous about it being possible to write to the database during a production job (just as you don't open a raw data file in "UPDATE" mode).  Every experiment I've worked on before has production jobs that can only treat the calibration database as a read-only resource (enforced by account permissions), and having been the guy who tidies up DB-writing accidents after experts have accidents, that makes a lot of sense. 

SQL query text is not very difficult to machine output.  I think maybe it's better just to output what we need into a text (`.sql`) file, then 'upload' a finished set to a database manually on the command line. This ensures:
- Modifications to the DB are fast, and - even better - done at known times.
- The constants get eye-balled by a human for sanity before uploading.
- Recognised pathologies can be manually edited by a human before uploading.
- We don't have to add complexity (and possible bugs) to a resource manager for what should be a rare task.

**Comment by jrquirk** [comment]
We've agreed to do this by outputting to a CSV file, doing spot checks, then using a script to update the database. Though enforcing this with file permissions never came up.

**Comment by jrquirk** [comment]
@AndrewEdmonds11 

I've updated the SetupNavigator so that it spits out a CSV file if the command line flag -c is passed to rootana. The CSV file that's output is the database that was loaded plus any additions/changes made by the modules to fPedestals, fNoises, or fCoarseTimeOffsets. This was the easiest way to do it quickly because we need the pedestals to get the cf time offsets. The files are calib.run#####.pedestals_and_noises.csv and calib.run#####.CoarseTimeOffset.csv, and they're saved into the working directory of the program.
If changes are made (like if the export_sql flag is true in CFTime generator) but the -c argument wasn't used, a warning is printed but because this happens at the end of the job usually, an exception isn't thrown.
I've made a pedestal/noise setter in SetupNavigator as I think it might be implemented.

I won't finish the CSV->sqlite DB script until tomorrow.

**Comment by AndrewEdmonds11** [comment]
Cheers, John. So it looks like the only thing left to do is get PlotTPI_PedestalAndNoise to use the `SetupNavigator::SetPedestalAndNoise` function and not write to the database.

Should we also standardise the names of the tables (i.e. change "pedestals_and_noises" to "PedestalAndNoise")? And also change the name of the SQLite file to calib.db or something?

**Comment by AndrewEdmonds11** [comment]
> This was the easiest way to do it quickly because we need the pedestals to get the cf time offsets.

Does this also mean that we have to do three runs in total then? One to get the pedestals and noises, then a second to get the time offsets and a third to do the final plots we want?

Since we should only need to do this once, it should be fine but thought it best to clarify

**Comment by AndrewEdmonds11** [comment]
> So it looks like the only thing left to do is get PlotTPI_PedestalAndNoise to use the `SetupNavigator::SetPedestalAndNoise` function and not write to the database.

Done and merged onto develop. I'll leave the decision to rename to you John since you're writing the CSV-->SQL script

**Comment by jrquirk** [comment]
I'll choose a consistent naming scheme.

> Does this also mean that we have to do three runs in total then?

Yeah.

**Comment by jrquirk** [comment]
I know this sounds like a silly question @AndrewEdmonds11, but in the calib db you have run number, channel (detector name), and bank name. Is it okay to drop the bank name?

**Comment by AndrewEdmonds11** [comment]
Sure. I can't remember why I put it in there...
